[JIT] Add support for with statements

SplitInfinity opened the pull request:
**Summary**
This commit adds support for with statements to PyTorch JIT. Each
of the with items in a with statement is represented in the JIT IR
as a pair of `prim::Enter` and `prim::Exit` nodes that call the
`__enter__` and `__exit__` methods defined on the context manager objects
returned by the expressions in the with item.

**Testing**
This commit adds unit tests for with statements with named with items,
nameless with items, and with statements that encounter exceptions.
```
$ python test/test_jit.py TestWith.test_with_as
Fail to import hypothesis in common_utils, tests are not derandomized
.
----------------------------------------------------------------------
Ran 1 test in 0.430s

OK
```

```
$ python test/test_jit.py TestWith.test_with_no_as
Fail to import hypothesis in common_utils, tests are not derandomized
.
----------------------------------------------------------------------
Ran 1 test in 0.264s

OK
```

```
$ python test/test_jit.py TestWith.test_with_exceptions
Fail to import hypothesis in common_utils, tests are not derandomized
Couldn't download test skip set, leaving all tests enabled...
.
----------------------------------------------------------------------
Ran 1 test in 1.053s

OK
```

Commit:
[JIT] Add support for with statements

Summary:
This commit adds support for with statements to PyTorch JIT. Each
of the with items in a with statement is represented in the JIT IR
as a pair of prim::Enter and prim::Exit nodes that call the
__enter__ and __exit__ methods defined on the context manager objects
returned by the expressions in the with item.

Testing:
This commit adds unit tests for with statements with named with items,
nameless with items, and with statements that encounter exceptions.

SplitInfinity commented:
Also, the tests pass if you run them one by one, but not if you run the whole suite:
```
$ python test/test_jit.py TestWith
Fail to import hypothesis in common_utils, tests are not derandomized
..Segmentation fault: 11
```
They never fail individually so I don't know what's up...

eellison reviewed (commented):
Just did a quick look through. 
Do you mind commenting with description of the high-level approach? What the IR looks like with this approach for a couple different example functions.

Maybe some other things to address would be how this works with nested with statements, and how this handles return / break / continue statements, which also could occur potentially within nested with statements. 

Obviously first PR may not handle all these cases, so maybe comment on what things we have to think about how to solve moving forward.

SplitInfinity commented:
### Frontend
The grammar for the `with statement` is:
```
with_stmt ::=  "with" with_item ("," with_item)* ":" suite
with_item ::=  expression ["as" target]
```
To parse this, this commit introduces three new lexer tokens: `TK_WITH (with)`, `TK_WITH_ITEM_NO_AS ()`, `TK_WITH_ITEM_AS (as)`. It also introduces four new `TreeViews`: `With` (extends `Stmt`), `WithItem`, `WithItemAs` (extends `WithItem`), and `WithItemNoAs` (extends `WithItem`):
```
With = (List<WithItem> withItems, List<Stmt> body)
WithItem = WithItemNoAs (Expr expression)
         | WithItemAs (Expr expression, Var target)
```
The parsing of a `with statement` is very similar to that of a `for loop`:
1) Parse a `TK_WITH` token, followed by a comma-delimited list of `WithItems`, followed by `:`, `TK_NEWLINE` and `TK_INDENT`.
2) Parse the body using pre-existing logic for parsing an indented block of statements.

Each with item is parsed by delegating to pre-existing logic for parsing expressions and identifiers, with some additional logic to check if the the expression is followed by a `TK_WITH_ITEM_AS` token (i.e. if the item is of the form `expression as target`).

### IR Emitter
A `With` `TreeView` is emitted as a series of pairs of `prim::Enter` and `prim::Exit` primitives surrounding the IR for its body. There is one `prim::Enter` and `prim::Exit` node for each with item. The `prim::Enter` nodes are emitted before the body of the `With` in the same order as the with items in the source code. The `prim::Exit` nodes are emitted after the body in the reverse order of the `prim::Enter` nodes. For example, the following Python code
```
with get_mgr() as a, get_other_mgr as b:
   <stmts>
```
would be emitted as the following IR
```
%1 : [type of get_mgr().__enter__] = prim::Enter(%a)
%2 : [type of get_other_mgr().__enter__] = prim::Enter(%b)
<stmts IR>
%3 : = prim::Exit(%b)
prim::drop(%3)
%4 : = prim::Exit(%a)
prim::drop(%4)
```
There is code to insert the `Exit` nodes before any early return statements, but I don't think it works in all cases (e.g. multiple early returns). `break` and `continue` are not handled.

### Interpreter
`prim::Enter` and `prim::Exit` IR nodes are emitted as `ENTER` and `EXIT` interpreter instructions. These work in the same way as `CALL` instructions, but with extra post processing steps. During instruction emission, the instruction numbers of all `EXIT` instructions are recorded in a list named `exits_` in the order that they appear in the code for a given function. During execution, there are two pointers in the `InterpreterState` that are updated as the code executes: `called` points to *one past the last `EXIT` that has already been called* and `should_call` points to *one past the last `EXIT` that should be called in the event of an exception*. Sounds confusing, but let me explain. If the code executes without a problem, the `EXIT` instructions should be called in the same order that they appear in `exits_`. Every time an `ENTER` instruction is executed, `should_call` is incremented, and every time an `EXIT` instruction is executed, `called` is incremented. If all goes well, they both equal `exits_.size()` when execution is complete.

However, suppose an exception is encountered after an `ENTER` instruction is called but before its corresponding `EXIT` is called. Then, `should_call` will be ahead of `called` by the number of objects that have been `__enter__`'d but not `__exit__`'d. Before dying, the interpreter goes through all frames in reverse order, and executes all `EXIT` instructions between the instruction numbers in `exits_` pointed to by `called` and `should_called`. In other words, it calls `__exit__` on all objects have been `__enter__`'d but not `__exit__`'d at the time of the exception. This is done by creating a new `Interpreter` instance to execute all the `__exit__` functions.

### Exporter
Exporting is pretty straightforward, but one edge case is an empty with-statement whose body consists of a `pass` statement. I checked for this by emitting a `pass` statement if the `prim::Enter` and `prim::Exit` are adjacent in the IR.

### Unit Tests
The current battery of tests includes:
- empty body (i.e. body consisting only of `pass`)
- early return
- two withs in a row
- two nested withs
- two combined withs (two with items)
- with that throws an exception
- nested with statements with an exception in the innermost with
- with statements in two different frames with an exception thrown in the second

### Known Issues
- need to handle `fork/wait`, `break`, `continue`, early `return`

eellison commented:
@SplitInfinity awesome, thank you for the comprehensive write up! This all looks really good. I'm going to throw out another option for how to represent with statements in the IR and we can think about the pros / cons in each approach.

The PEP for with statements describe them as desugaring to:
```
VAR = EXPR
VAR.__enter__()
try:
    BLOCK
finally:
    VAR.__exit__()
```

We could represent with statements as a node with two blocks, one representing the `try` block and one representing the `finally` block. 

```
def test_serial(x):
    # type: (Tensor) -> Tensor

    c = Context(1)

    with c as mult:
        y = x ** mult

    return y
```
would compile to something like:
```
  %y : Tensor = prim::With(%cond.1) # test/test_jit.py:11863:12
    block0():
      %y.1 : Tensor = aten::pow(%x.1, %mult.1) # test/test_jit.py:11864:20
      -> (%g.1)
    block1():
      %3 : None = prim::CallMethod[name="__exit__"](%c.1, %1)
      -> (%g.2)
  return (%y)
```

A `with` statement within a `with` statement would have another `with` node within block0. Similarly, a statement like:
```
with A() as a, A() as b:
            pass
```
desugars to a nested within statement. 

```
  %y : Tensor = prim::With(%cond.1) # test/test_jit.py:11863:12
    block0():
      %y.1 : Tensor = aten::pow(%x.1, %mult.1) # test/test_jit.py:11864:20
      % .... = prim::With(%cond.1):
           block0():
                ...
           block1():
               ...
      -> (%g.1)
    block1():
      %3 : None = prim::CallMethod[name="__exit__"](%c.1, %1)
      -> (%g.2)
  return (%y)
```

A couple advantages of this approach:
- it makes it easy to implement with early returns, break, continue. The `exit_transform` pass operates can just ignore the second block with statements. Whenever a `prim::With` node executes, the second block will always execute, early returns / breaks / continues get lowered out of the graph, and the second block always get executed. 
- More closely maps to python bytecode, also if we ever have `try ... finally` this IR would easily map to that. 
- All method calls can be inlined. When we don't inline method calls it prohibits optimization and also (maybe) makes it harder to export to mobile. This could also be addressed by adding a block to `prim::Exit`. 


The main thing there is that i'm not sure how `prim::Exit` nodes would work with the early returns/continues/breaks/continues, but it's easy to implement with the block approach.

What are your thoughts ?

SplitInfinity commented:
```
  %y : Tensor = prim::With(%cond.1) # test/test_jit.py:11863:12
    block0():
      %y.1 : Tensor = aten::pow(%x.1, %mult.1) # test/test_jit.py:11864:20
      -> (%g.1)
    block1():
      %3 : None = prim::CallMethod[name="__exit__"](%c.1, %1)
      -> (%g.2)
  return (%y)
```

Where is the call to `__enter__`? Is it right before the `prim::With` node? If not, what defines `%mult.1` so that it has a value in `block0`? And what sets `%y` correctly so that the returned value is correct?

eellison commented:
@SplitInfinity yes it's right before the `__enter__` call. This is consistent with python bytecode in that if an exception is thrown in the `__enter__` call `__exit__` does not get invoked
```
VAR = EXPR
VAR.__enter__()
try:
    BLOCK
finally:
    VAR.__exit__()
```

SplitInfinity commented:
Took a while but I just pushed a version that adds `With` blocks before doing any exit transformations and removes them after the pass is done. The basic idea is as follows:
- for every `Enter` and `Exit` pair, add `With` blocks that contain all nodes between them in `block0` and just the `Exit` node in `block1`, and insert the `With` block after the `Enter`
- do exit transformations as usual; with blocks don't introduce any extra control flow, so the correct outputs for `block0` are whatever `transformExits(block0)` returns. We know `block1` will always be executed, so we add uninitialized output values of the same types as the ones for `block0` to keep the IR sane, knowing that these values will never be used
- remove each `With` block by adding all of the nodes in `block0` after the `Enter` preceding the `With` and then the singular `Exit` in `block1`

eellison commented:
> Took a while but I just pushed a version that adds `With` blocks before doing any exit transformations and removes them after the pass is done. The basic idea is as follows:
> 
> * for every `Enter` and `Exit` pair, add `With` blocks that contain all nodes between them in `block0` and just the `Exit` node in `block1`, and insert the `With` block after the `Enter`
> * do exit transformations as usual; with blocks don't introduce any extra control flow, so the correct outputs for `block0` are whatever `transformExits(block0)` returns. We know `block1` will always be executed, so we add uninitialized output values of the same types as the ones for `block0` to keep the IR sane, knowing that these values will never be used
> * remove each `With` block by adding all of the nodes in `block0` after the `Enter` preceding the `With` and then the singular `Exit` in `block1`
> 
> I added some extra unit tests that use `break` and `continue` inside an `if` and they seem to work. However, while I can run the tests individually (i.e. `python test/test_jit.py TestWith.test_with_as`), running all of them at once (i.e. `python test/test_jit.py TestWith`) segfaults. Any advice on how to debug this?

Where is the segfault occuring ?

SplitInfinity commented:
Here's the full stacktrace, I'm still trying to make sense of it myself:
```
(base) meghanl-mbp:pytorch meghanl$ python test/test_jit.py TestWith
Fail to import hypothesis in common_utils, tests are not derandomized
..Segmentation fault: 11
(base) meghanl-mbp:pytorch meghanl$ lldb -- python test/test_jit.py TestWith
(lldb) target create "python"
Current executable set to 'python' (x86_64).
(lldb) settings set -- target.run-args  "test/test_jit.py" "TestWith"
(lldb) r
Process 44468 launched: '/opt/anaconda3/bin/python' (x86_64)
Fail to import hypothesis in common_utils, tests are not derandomized
..Process 44468 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
    frame #0: 0x00000001063f1c8b libtorch_python.dylib`void pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >::init_holder<c10::Type>(pybind11::detail::instance*, pybind11::detail::value_and_holder&, std::__1::shared_ptr<c10::ClassType> const*, std::__1::enable_shared_from_this<c10::Type> const*) + 27
libtorch_python.dylib`pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >::init_holder<c10::Type>:
->  0x1063f1c8b <+27>: movq   0x10(%rax), %rdi
    0x1063f1c8f <+31>: testq  %rdi, %rdi
    0x1063f1c92 <+34>: je     0x1063f1ee0               ; <+624>
    0x1063f1c98 <+40>: movq   0x8(%rax), %r12
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x10)
  * frame #0: 0x00000001063f1c8b libtorch_python.dylib`void pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >::init_holder<c10::Type>(pybind11::detail::instance*, pybind11::detail::value_and_holder&, std::__1::shared_ptr<c10::ClassType> const*, std::__1::enable_shared_from_this<c10::Type> const*) + 27
    frame #1: 0x00000001063f1baf libtorch_python.dylib`pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >::init_instance(pybind11::detail::instance*, void const*) + 239
    frame #2: 0x00000001063f2109 libtorch_python.dylib`void pybind11::cpp_function::initialize<void pybind11::detail::initimpl::factory<torch::jit::initPythonIRBindings(_object*)::$_91, pybind11::detail::void_type (*)(), std::__1::shared_ptr<c10::ClassType> (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&), pybind11::detail::void_type ()>::execute<pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> > >(pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >&) &&::'lambda'(pybind11::detail::value_and_holder&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&), void, pybind11::detail::value_and_holder&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::detail::is_new_style_constructor>(pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >&&,  (*)(void pybind11::detail::initimpl::factory<torch::jit::initPythonIRBindings(_object*)::$_91, pybind11::detail::void_type (*)(), std::__1::shared_ptr<c10::ClassType> (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&), pybind11::detail::void_type ()>::execute<pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> > >(pybind11::class_<c10::ClassType, c10::Type, std::__1::shared_ptr<c10::ClassType> >&) &&::'lambda'(pybind11::detail::value_and_holder&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)...), void const&...)::'lambda'(pybind11::detail::function_call&)::__invoke(pybind11::detail::function_call&) + 377
    frame #3: 0x0000000105e70d31 libtorch_python.dylib`pybind11::cpp_function::dispatcher(_object*, _object*, _object*) + 3729
    frame #4: 0x000000010002e603 python`_PyMethodDef_RawFastCallDict + 483
    frame #5: 0x000000010002dd5f python`_PyObject_FastCallDict + 111
    frame #6: 0x0000000100032002 python`method_call + 130
    frame #7: 0x000000010002fa82 python`PyObject_Call + 130
    frame #8: 0x00000001000b04b6 python`slot_tp_init + 422
    frame #9: 0x00000001000ba4d1 python`type_call + 241
    frame #10: 0x000000010002ee03 python`_PyObject_FastCallKeywords + 179
    frame #11: 0x000000010016bd75 python`call_function + 453
    frame #12: 0x00000001001699c5 python`_PyEval_EvalFrameDefault + 46149
    frame #13: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #14: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #15: 0x0000000106432d54 libtorch_python.dylib`pybind11::object pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::str_attr> >::operator()<(pybind11::return_value_policy)1, pybind11::object&, torch::jit::SourceRange const&>(pybind11::object&, torch::jit::SourceRange const&) const + 52
    frame #16: 0x0000000106432b08 libtorch_python.dylib`torch::jit::(anonymous namespace)::PythonResolver::resolveType(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, torch::jit::SourceRange const&) + 408
    frame #17: 0x000000014999f123 libtorch_cpu.dylib`torch::jit::FunctionResolver::resolveType(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, torch::jit::SourceRange const&) + 19
    frame #18: 0x00000001499b7a34 libtorch_cpu.dylib`torch::jit::Environment::getSugaredVar(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, torch::jit::SourceRange const&, bool) + 212
    frame #19: 0x00000001499aa1ca libtorch_cpu.dylib`torch::jit::to_ir::emitSugaredExpr(torch::jit::Expr const&, unsigned long, std::__1::shared_ptr<c10::Type> const&) + 378
    frame #20: 0x00000001499d7168 libtorch_cpu.dylib`torch::jit::to_ir::emitApplyExpr(torch::jit::Apply&, unsigned long, std::__1::shared_ptr<c10::Type> const&) + 104
    frame #21: 0x00000001499aa0f8 libtorch_cpu.dylib`torch::jit::to_ir::emitSugaredExpr(torch::jit::Expr const&, unsigned long, std::__1::shared_ptr<c10::Type> const&) + 168
    frame #22: 0x00000001499cca3c libtorch_cpu.dylib`torch::jit::to_ir::emitSingleAssignment(torch::jit::Assign const&) + 1100
    frame #23: 0x00000001499a8cb5 libtorch_cpu.dylib`torch::jit::to_ir::emitAssignment(torch::jit::Assign const&) + 181
    frame #24: 0x00000001499a3658 libtorch_cpu.dylib`torch::jit::to_ir::emitStatements(torch::jit::ListIterator<torch::jit::Stmt>, torch::jit::ListIterator<torch::jit::Stmt>) + 1688
    frame #25: 0x00000001499a0842 libtorch_cpu.dylib`torch::jit::to_ir::emitDef(torch::jit::Def const&, torch::jit::Self const*, torch::jit::Block*) + 290
    frame #26: 0x000000014999fdff libtorch_cpu.dylib`torch::jit::to_ir::to_ir(torch::jit::Def const&, std::__1::shared_ptr<torch::jit::Resolver>, torch::jit::Self const*, torch::jit::Function&) + 655
    frame #27: 0x000000014999f8b5 libtorch_cpu.dylib`std::__1::__function::__func<torch::jit::CompilationUnit::define(c10::optional<c10::QualifiedName> const&, torch::jit::Def const&, std::__1::shared_ptr<torch::jit::Resolver> const&, torch::jit::Self const*, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, torch::jit::Function*, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, torch::jit::Function*> > > const&, bool) const::$_0, std::__1::allocator<torch::jit::CompilationUnit::define(c10::optional<c10::QualifiedName> const&, torch::jit::Def const&, std::__1::shared_ptr<torch::jit::Resolver> const&, torch::jit::Self const*, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, torch::jit::Function*, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, torch::jit::Function*> > > const&, bool) const::$_0>, void (torch::jit::GraphFunction&)>::operator()(torch::jit::GraphFunction&) + 533
    frame #28: 0x0000000149a8728b libtorch_cpu.dylib`torch::jit::GraphFunction::ensure_defined() + 171
    frame #29: 0x000000014999bfb9 libtorch_cpu.dylib`torch::jit::CompilationUnit::define(c10::optional<c10::QualifiedName> const&, std::__1::vector<torch::jit::Def, std::__1::allocator<torch::jit::Def> > const&, std::__1::vector<std::__1::shared_ptr<torch::jit::Resolver>, std::__1::allocator<std::__1::shared_ptr<torch::jit::Resolver> > > const&, torch::jit::Self const*, bool) + 713
    frame #30: 0x000000014999a8eb libtorch_cpu.dylib`torch::jit::CompilationUnit::define(c10::optional<c10::QualifiedName> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<torch::jit::Resolver> const&, torch::jit::Self const*) + 571
    frame #31: 0x000000010643f392 libtorch_python.dylib`void pybind11::cpp_function::initialize<torch::jit::initJitScriptBindings(_object*)::$_21, void, torch::jit::CompilationUnit&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<pybind11::function (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)>, pybind11::name, pybind11::is_method, pybind11::sibling>(torch::jit::initJitScriptBindings(_object*)::$_21&&, void (*)(torch::jit::CompilationUnit&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::function<pybind11::function (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)>), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&)::'lambda'(pybind11::detail::function_call&)::__invoke(pybind11::detail::function_call&) + 562
    frame #32: 0x0000000105e70d31 libtorch_python.dylib`pybind11::cpp_function::dispatcher(_object*, _object*, _object*) + 3729
    frame #33: 0x000000010002f4e8 python`_PyMethodDef_RawFastCallKeywords + 392
    frame #34: 0x000000010002efa0 python`_PyObject_FastCallKeywords + 592
    frame #35: 0x000000010016bd75 python`call_function + 453
    frame #36: 0x0000000100168a83 python`_PyEval_EvalFrameDefault + 42243
    frame #37: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #38: 0x000000010002f313 python`_PyFunction_FastCallKeywords + 195
    frame #39: 0x000000010016bc67 python`call_function + 183
    frame #40: 0x0000000100169a7e python`_PyEval_EvalFrameDefault + 46334
    frame #41: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #42: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #43: 0x00000001000b03d1 python`slot_tp_init + 193
    frame #44: 0x00000001000ba4d1 python`type_call + 241
    frame #45: 0x000000010002ee03 python`_PyObject_FastCallKeywords + 179
    frame #46: 0x000000010016bd75 python`call_function + 453
    frame #47: 0x0000000100169a7e python`_PyEval_EvalFrameDefault + 46334
    frame #48: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #49: 0x000000010002f313 python`_PyFunction_FastCallKeywords + 195
    frame #50: 0x000000010016bc67 python`call_function + 183
    frame #51: 0x0000000100169a7e python`_PyEval_EvalFrameDefault + 46334
    frame #52: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #53: 0x000000010002f313 python`_PyFunction_FastCallKeywords + 195
    frame #54: 0x000000010016bc67 python`call_function + 183
    frame #55: 0x00000001001689ed python`_PyEval_EvalFrameDefault + 42093
    frame #56: 0x000000010002ebe5 python`function_code_fastcall + 117
    frame #57: 0x000000010016bc67 python`call_function + 183
    frame #58: 0x00000001001699c5 python`_PyEval_EvalFrameDefault + 46149
    frame #59: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #60: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #61: 0x0000000100032002 python`method_call + 130
    frame #62: 0x000000010002fa82 python`PyObject_Call + 130
    frame #63: 0x0000000100169bf2 python`_PyEval_EvalFrameDefault + 46706
    frame #64: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #65: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #66: 0x00000001000ae4bd python`slot_tp_call + 189
    frame #67: 0x000000010002ee03 python`_PyObject_FastCallKeywords + 179
    frame #68: 0x000000010016bd75 python`call_function + 453
    frame #69: 0x00000001001699c5 python`_PyEval_EvalFrameDefault + 46149
    frame #70: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #71: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #72: 0x0000000100032002 python`method_call + 130
    frame #73: 0x000000010002fa82 python`PyObject_Call + 130
    frame #74: 0x0000000100169bf2 python`_PyEval_EvalFrameDefault + 46706
    frame #75: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #76: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #77: 0x00000001000ae4bd python`slot_tp_call + 189
    frame #78: 0x000000010002ee03 python`_PyObject_FastCallKeywords + 179
    frame #79: 0x000000010016bd75 python`call_function + 453
    frame #80: 0x00000001001699c5 python`_PyEval_EvalFrameDefault + 46149
    frame #81: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #82: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #83: 0x0000000100032002 python`method_call + 130
    frame #84: 0x000000010002fa82 python`PyObject_Call + 130
    frame #85: 0x0000000100169bf2 python`_PyEval_EvalFrameDefault + 46706
    frame #86: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #87: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #88: 0x00000001000ae4bd python`slot_tp_call + 189
    frame #89: 0x000000010002ee03 python`_PyObject_FastCallKeywords + 179
    frame #90: 0x000000010016bd75 python`call_function + 453
    frame #91: 0x00000001001699c5 python`_PyEval_EvalFrameDefault + 46149
    frame #92: 0x000000010002ebe5 python`function_code_fastcall + 117
    frame #93: 0x000000010016bc67 python`call_function + 183
    frame #94: 0x00000001001689ed python`_PyEval_EvalFrameDefault + 42093
    frame #95: 0x000000010002ebe5 python`function_code_fastcall + 117
    frame #96: 0x000000010016bc67 python`call_function + 183
    frame #97: 0x00000001001689ed python`_PyEval_EvalFrameDefault + 42093
    frame #98: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #99: 0x000000010002e0d7 python`_PyFunction_FastCallDict + 231
    frame #100: 0x00000001000b03d1 python`slot_tp_init + 193
    frame #101: 0x00000001000ba4d1 python`type_call + 241
    frame #102: 0x000000010002ee03 python`_PyObject_FastCallKeywords + 179
    frame #103: 0x000000010016bd75 python`call_function + 453
    frame #104: 0x0000000100169a7e python`_PyEval_EvalFrameDefault + 46334
    frame #105: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #106: 0x000000010002f313 python`_PyFunction_FastCallKeywords + 195
    frame #107: 0x000000010016bc67 python`call_function + 183
    frame #108: 0x00000001001699c5 python`_PyEval_EvalFrameDefault + 46149
    frame #109: 0x000000010015d34a python`_PyEval_EvalCodeWithName + 410
    frame #110: 0x00000001001c0b80 python`PyRun_FileExFlags + 256
    frame #111: 0x00000001001bfff7 python`PyRun_SimpleFileExFlags + 391
    frame #112: 0x00000001001edccf python`pymain_main + 9711
    frame #113: 0x00000001000018dd python`main + 125
    frame #114: 0x00007fff688583d5 libdyld.dylib`start + 1
```

eellison commented:
I would guess some sort of bad access but that's not very helpful. after I review and you make changes i can check it out locally and try to debug.

eellison commented on torch/csrc/jit/ir/ir.cpp:
```diff
@@ -1048,6 +1048,8 @@ bool Node::hasSideEffects() const {
     case prim::Guard:
     case prim::rpc_async: // It represents RPC message sent.
     case aten::wait: // It can represent RPC message received.
+    case prim::Enter:
```
Can you add description of your PRs changes where applicable to  https://github.com/pytorch/pytorch/blob/f4532d7542cc819786af9a3feb747be60d3de2b9/torch/csrc/jit/OVERVIEW.md ?

and also edit https://github.com/pytorch/pytorch/blob/master/docs/source/jit_python_reference.rst to say we support with statements now

SplitInfinity replied:
Done. Although I would be more comfortable doing this after more testing or at least making sure `torch.no_grad` works.

eellison replied:
Yea, good point. up to you

eellison commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -376,6 +376,11 @@ struct CodeImpl {
+  // An ordered list of instruction numbers for EXIT instructions. This is
+  // used to invoke __exit__ on objects that have been __enter__'d in the right
+  // order when an exception has been encountered.
+  std::vector<size_t> exits_;
```
@zdevito do you mind reviewing the interpreter changes ?

eellison reviewed (commented):
Part one of review: just reviewed the exit transform pass. 

Looks great! I have one small suggestion that I think might make the pass easier to reason about.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1733,6 +1736,82 @@ struct to_ir {
+        }
+
+        // Set the output of the enter node to be the return type of __enter__.
+        n->output(0)->setType(rhsClass->getMethod("__enter__")
```
I think we should be doing some checking of the `__enter__` and `__exit__` schema to check that they are of an expected form.

SplitInfinity replied:
Here are the schemas from one of my tests:
```
__enter__(__torch__.jit.test_with.Context self) -> (int)
__exit__(__torch__.jit.test_with.Context self, Tensor type, Tensor value, Tensor tb) -> (None)
```
For both methods, the first arg must be `self` because these are both class methods, and they are fetched from the `ClassType` object corresponding to the class they belong to. So checking that the first arg is `self` seems pointless.

In the case of `__enter__`, the return type is `int` because that's how I wrote the example. That' s not always the case. 

In the case of `__exit__`, we don't actually use the last three args, but they have to be present. So I think it would be good to check that `__exit__` takes 4 args.

eellison reviewed (commented):
Reviewing the other parts: looks great, a few small comments. 

How exactly are you handling the extra arguments in `__enter__` and `__exit__` ? What values do we pass to `__exit__` and what are the type signatures that should succeed in `__enter__` / `__exit__`? 

We should also try to figure out the next steps for with statements and document them. (Which are not needed for this PR). A potential few I can think of: 
- add support for `torch.no_grad`, which is the most typical with usage in pytorch code (and then edit back in [this](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1674)
- context managers which are registered in a specific way could be used as decorators, because:
`@torch.no_grad` is a very common pattern
- optimization to remove `with no_grad` blocks when the module is frozen or we've profiled that all inputs to the block don't require grad.

SplitInfinity commented on torch/csrc/jit/frontend/exit_transforms.cpp:
```diff
@@ -303,6 +303,168 @@ struct ExitTransformer {
+    std::vector<Node*> stack;
+
+    for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
+      if (it->kind() == prim::Enter && it->next()->kind() != prim::With) {
```
It was needed because I was doing the transformation to/from the block-based representation in `transformExits`, which gets called recursively on the body block of a `prim::With` and might end up processing `prim::With`. But it's no longer needed because I moved the transformation of the `Enter/Exit` pairs to and from `With` nodes to the location you suggested so that we only do those transforms once.

SplitInfinity commented on torch/csrc/jit/frontend/exit_transforms.cpp:
```diff
@@ -303,6 +303,168 @@ struct ExitTransformer {
+
+      // Recursively remove With nodes in the body block. The exit block
+      // contains only an Exit node so it does not need to be considered here.
+      removeWithNodes(body_block);
```
Good point, done.

SplitInfinity commented on torch/csrc/jit/frontend/exit_transforms.cpp:
```diff
@@ -303,6 +303,168 @@ struct ExitTransformer {
+
+    // The outputs of the body block are whatever transformExits found after
+    // analyzing it.
+    registerBlockOutputs(body_block, body_pair.exitValues());
```
Hm, but then what propagates these outputs back up to the containing scope?

eellison replied:
There are two options here:
- temporarily move the nodes between `prim::Enter` and `prim::Exit` into the with block, leave values unchanged

- move nodes between `prim::Enter` and `prim::Exit` into the with block, and calculate all values that now defined in the `prim::With` block and that have uses in the rest of the graph. For those values, add them as outputs of the `prim::With` node and then replace their uses with the `prim::With` outputs. Then, when you inline `prim::With` node, thread back the correct uses.
```
  %mult.1 : Tensor = prim::Enter(%c.1)
  %y.1 : Tensor = aten::pow(%x.1, %mult.1) # /Users/meghanl/Desktop/pytorch/test/jit/test_with.py:104:20
  %8 : Tensor = prim::Exit(%c.1)
  % = prim::Print(%y.1)
```
-> 
```
% y.2 : Tensor = prim::With(%c.1):
    block0():
      %y.1 : Tensor = aten::pow(%x.1, %mult.1)
      -> (%g.1)
    block1():
           ....
% = prim::Print(%y.2)
```

We can go either approach. But right now we're only adding block outputs for the exit values and none of the other values defined in the block, and we're not having them used anywhere. I personally think the first approach is easier to reason about and less logic to maintain.

SplitInfinity commented on torch/csrc/jit/frontend/tree_views.h:
```diff
@@ -954,6 +959,83 @@ struct Var : public Expr {
 
+// WithItem represents an item using with a WithStmt. It can be a WithItemNoAs
+// or a WithItemAs.
+struct WithItem : public TreeView {
```
Done.

SplitInfinity commented on torch/jit/frontend.py:
```diff
@@ -125,6 +125,11 @@ class FrontendTypeError(FrontendError):
 
+def build_withitems(ctx, items):
+    items = [build_withitem(ctx, i) for i in items]
+    return list(filter(None, items))
```
I saw it being used for `build_stmt` below but it turns out it is not required.

SplitInfinity commented on torch/jit/frontend.py:
```diff
@@ -265,6 +270,25 @@ def get_default_args(fn):
+        end = item.context_expr.col_offset + len(item.context_expr.id)
+
+        if item.optional_vars:
+            end += len(" as ")
```
Good point, fixed.

SplitInfinity commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1733,6 +1736,82 @@ struct to_ir {
+    Block* current_block = nullptr;
+
+    for (const auto& target : targets) {
+      if (target.kind() == TK_WITH_ITEM_AS) {
```
Done.

SplitInfinity commented:
Pushed an update with the following changes:
* Addressed comments to the changes to the `exit_transform` pass
  * Moved the insertion/removal of `prim::With` nodes to `jit::TransformExits` so that they are each performed only once
  * Modified insert/removal code to recurse into `prim::With`, `prim::If` and `prim::Loop` nodes while looking for `prim::Enter`/`prim::Exit` (in the case of insertion) and `prim::With` (in the case of removal) nodes. Previously, the code was only recursing into `prim::With` nodes which is not correct. I essentially had to implement a depth-first iterator for `Graph` to do this, and I may have reinvented the wheel in doing so.
* Consolidated `WithItemAs` and `WithItemNoAs` into one `WithItem` `TreeView` class
* Updated `OVERVIEW.md` and `jit_python_reference.rst`
* Addressed nits

eellison commented on torch/csrc/jit/frontend/exit_transforms.cpp:
```diff
@@ -505,6 +529,260 @@ struct ExitTransformer {
+    // Record all nodes that need to be appended after the Enter that precedes
+    // the With block to avoid iterator invalidation issues later when moving
+    // nodes around.
+    for (auto it = body_block->nodes().begin(); it != body_block->nodes().end();
```
i forget to do this too but a nit is just to use `(for auto node: block->nodes())` if you're not doing any invalidation while you're iterating

eellison commented on torch/csrc/jit/frontend/lexer.h:
```diff
@@ -106,7 +106,10 @@ namespace jit {
+  _(TK_IMPORT, "import", "import")               \
+  _(TK_WITH, "with", "with")                     \
+  _(TK_WITH_ITEM, "withitem", "")                \
+  _(TK_WITH_ITEM_AS, "withitemas", "as")
```
I don't know if we ever want to support importing in TorchScript or catching exceptions.. if we do `as` can also be used as `import x as y` + `catch Exception as exc`. maybe rename `TK_WITH_ITEM_AS` to be less with statement specific?

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -460,6 +462,7 @@ void AliasDb::analyzeImpl(Node* node) {
       return;
     case prim::Print:
     case prim::isinstance:
+    case prim::With:
```
As with `prim::Load` and `prim::Store`, `prim::With` is an ir construct that only exists before the IR is in SSA form and valid for Alias Analysis to be called on it. If someone does it should be a hard error, so I think we want to remove this.

eellison commented on torch/csrc/jit/runtime/operator.cpp:
```diff
@@ -277,6 +277,9 @@ bool aliasAnalysisHasSpecialCaseFor(Symbol symbol) {
       prim::rpc_async,
+      prim::Enter,
+      prim::Exit,
+      prim::With,
```
ditto to above to remove `prim::With` here

eellison commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -739,6 +739,27 @@ struct PythonPrintImpl {
+        }
+        level--;
+      } break;
       case prim::Function: {
```
Can you add `as` and `with` to the list of `reserved_names`? i don't think this is BC-breaking bc in python right now if you try to assign a variable to either you'll get a hard error
```
>>> as = 4
  File "<stdin>", line 1
    as = 4
SyntaxError: invalid syntax
>>> with = 5
  File "<stdin>", line 1
    with = 5
         ^
SyntaxError: invalid syntax
```

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1733,6 +1736,57 @@ struct to_ir {
+      }
+
+      // Set the output of the enter node to be the return type of __enter__.
+      n->output(0)->setType(
```
We should figure out what we're going to do about typing here. For `__enter__`, that's easy, we can check that it's a method with no arguments (although in practice no one will break that). For `__exit__` i'm not sure exactly what type signatures we should accept, but we should probably be checking something.

eellison commented on test/jit/test_with.py:
```diff
@@ -0,0 +1,387 @@
+            as expected.
+            """
+            with c:
+                y = x + c.count
```
could you add a test for when this conditionally early returns ?

eellison commented on test/jit/test_with.py:
```diff
@@ -0,0 +1,387 @@
+        test_list = [2, 0, 1, 3, 0, 2]
+
+        self.checkScript(test_basic, (test_input,))
+        # self.checkScript(test_pass, (test_input,))
```
These are commented out, probably on accident

eellison reviewed (commented):
Looks great!! I think this is pretty much ready to land we just need to figure out what we're going to do with the `__exit__` type signature. I'm not sure exactly...

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1121,6 +1159,45 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+                    ->get_executor()
+                    .getPlanFor(stack, GraphExecutor::getDefaultNumBailOuts())
+                    .code;
+            frames.back().pc = af.pc + 1;
```
The object itself is not being passed here.

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1121,6 +1159,45 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
               af.pc += inst.X;
             }
           } break;
+          case ENTER: {
```
This is pretty repetitive (and a copy/paste) of the CALL node. There are also some issues with ordering. For instance, what happens if an exception occurs in the call to enter? the object is already on the the __enter__/__exit__ stack, but the contract only says that __exit__ will be called _after_ enter.

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1345,10 +1422,47 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
     } catch (std::exception& e) {
       frames.back().pc = af.pc;
+
+      // Before dying, go through all the frames in reverse order and call
```
The amount of comments here suggest this logic is pretty complicated.

zdevito reviewed (commented):
This is only a review of the interpreter changes, I didn't look at anything else.

I think the current interpreter changes need some cleanup, otherwise they are going to make it hard to maintain. In particular the issues I see are:

* several copy/pastes of function invocation code, which will complicate future changes to how executors work
* lots of indexing and tracing of pc's for exit instructions that will complicate things.

Here is my suggestion, which is a small refactoring of this. Have two instructions:

* ENTER - pops the top of the stack, and pushes it onto enteredObjects. It does not invoke anything, this will be handled by a normal CALL instruction.
* EXIT - pops the enteredObject off the stack, and calls __exit__ on it. We should refactor the code to set up the interpreter to make a call so both CALL and EXIT share the enterFrame/get_executor/ActiveFrame logic.


To handle prim::enter, we then first emit a CALL_METHOD instruction on object, and then emit a ENTER after it. This ensures that the object stack doesn't prematurely contain the object before enter finishes.

To handle prim::exit, we simply emit the raw EXIT instruction (which itself takes no arguments).

On an except, simply loop in reverse order and call exit on the objects. You don't need to manage the sub-interpreter. code state here, just use the public API to call a method on an ivalue.

eellison commented on torch/csrc/jit/frontend/exit_transforms.cpp:
```diff
@@ -505,6 +529,260 @@ struct ExitTransformer {
+  // A stack of all blocks that need to be revisited when the current block has
+  // been processed, as well as the corresponding nodes that should be returned
+  // when those blocks are revisited. Think of it as the standard DFS stack.
+  std::vector<BlockIteratorPair> block_stack_;
```
This isn't necessary for this PR, maybe something to be when we start using this in other places.

SplitInfinity commented:
Addressed @eellison 's comments in a new commit (except for the ones with `DepthFirstGraphIterator`, I'll do that in a follow up or if we ever use it elsewhere). I'm gonna go back and make the interpreter changes now.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1733,6 +1737,83 @@ struct to_ir {
+        for (unsigned i = 1; i < 4; ++i) {
+          if (exitSchema.arguments().at(i).type() != AnyType::get()) {
+            throw ErrorReport(e.range())
+                << "argument " << i << " of __exit__ must have Any type";
```
can you add some context to this error message ? it won't clear to a user why we require `Any` type so pretty much everyone will hit this. Maybe something like:

```argument i of __exit__ must have Any type. TorchScript does not currently support passing type, value, or traceback to the __exit__ function.```

or something else as you see fit.

eellison commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1733,6 +1737,83 @@ struct to_ir {
+            << "__exit__ must have four arguments and no return value";
+      } else {
+        if (exitSchema.returns().at(0).type() != NoneType::get()) {
+          throw ErrorReport(e.range()) << "__exit__ must have no return value";
```
can u add a test or two for the failure conditions ?

eellison reviewed (approved):
Accept from me. Need an accept from @zdevito on the interpreter changes

SplitInfinity commented:
@zdevito 

I have made the interpreter changes you suggested, but ran into one small issue. Emitting two instructions for a `prim::Enter` (one `CALL` and one `ENTER`) results in the object that is being used in the `with statement` being `MOVE`d earlier than it should. This is because the decision to move a value is based on the number of uses in the IR, not in the emitted stream of instructions. I temporarily disabled all moves so that I could upload the rest of my changes and discuss with you what to do about this issue. Once we agree on the right way to address this, I will undo that change.

I think what makes the most sense is to emit a `prim::MethodCall` before the `prim::Enter`. What do you think?

SplitInfinity commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -526,7 +526,7 @@ struct CodeImpl {
     } else {
       int reg = registerFor(input);
-      bool moved = input->uses().size() == ++use_count_[input];
+      bool moved = false; // input->uses().size() == ++use_count_[input];
```
This is the temporary change I was referring to in my comment. I do not intend to land this, but the rest of my changes don't work without it for the reason listed in my top-level comment.

SplitInfinity commented:
Ping @zdevito

ilia-cher commented:
Thanks for working on this! This will be very useful for the profiler (e.g. `with record_function("...")`) and we've been already asked by the external users to support this. Would be great to get further reviews from the TorchScript team. cc. @suo @zdevito @jamesr66a

SplitInfinity commented:
FYI build failure is due to some ongoing work with profiling executor and alias analysis - I need #39497 to land before those will pass.

zdevito reviewed (approved):
Looks good with the one discussed fix.
